library/psa_crypto.c

```diff
@@ -2818,7 +2932,28 @@ static psa_status_t psa_mac_setup( psa_mac_operation_t *operation,
 
     if( truncated == 0 )
     {
-        /* The "normal" case: untruncated algorithm. Nothing to do. */
+        /* The "normal" case: untruncated algorithm. Re-validate the
+         * key policy with explicit MAC length set in the algorithm
+         * when the algorithm policy is at-least-this-length to catch
+         * a corner case due to the default MAC length being unknown
+         * at key loading time. */
+        if( PSA_ALG_IS_MAC( slot->attr.policy.alg ) &&
+            ( PSA_ALG_FULL_LENGTH_MAC( slot->attr.policy.alg ) == full_length_alg ) &&
+            ( slot->attr.policy.alg & PSA_ALG_MAC_AT_LEAST_THIS_LENGTH_FLAG ) )
+        {
+            /* validate policy length */
+            if( PSA_MAC_TRUNCATED_LENGTH( slot->attr.policy.alg ) > operation->mac_size )
+                status = PSA_ERROR_NOT_PERMITTED;
+        }
+
+        if( PSA_ALG_IS_MAC( slot->attr.policy.alg2 ) &&
```

Comment:
If `alg` and `alg2` are both at-least-length for the same MAC algorithm, and `alg` is that mac algorithm, the operation is allowed only if both policies allow it. It should in fact be allowed if at least one policy allows it.

This is a corner case that falls outside of expected usage, and it causes authorization to fail rather than succeed, so it isn't an absolute blocker in itself. However, the combination of this bug and of the general weakness of doing policy checks in an operation function, I judge that it isn't acceptable for maintainability.

Reply:
Updated